lnd_test.go

```diff
@@ -1464,6 +1468,150 @@ func testMultiHopPayments(net *networkHarness, t *harnessTest) {
+
+	// Open a channel with 100k satoshis between Alice and Bob with Alice
+	// being the sole funder of the channel.
+	ctxt, _ := context.WithTimeout(ctxb, timeout)
+	chanPointAlice := openChannelAndAssert(ctxt, t, net, net.Alice,
+		net.Bob, chanAmt, 0)
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	err := net.Alice.WaitForNetworkChannelOpen(ctxt, chanPointAlice)
+	if err != nil {
+		t.Fatalf("alice didn't advertise her channel: %v", err)
+	}
+
+	// Create a new node (Carol), load her with some funds, then establish
+	// a connection between Carol and Alice with a channel that has
+	// identical capacity to the one created above.
+	//
+	// The network topology should now look like: Carol -> Alice -> Bob
+	carol, err := net.NewNode(nil)
+	if err != nil {
+		t.Fatalf("unable to create new nodes: %v", err)
+	}
+	if err := net.ConnectNodes(ctxb, carol, net.Alice); err != nil {
+		t.Fatalf("unable to connect carol to alice: %v", err)
+	}
+	err = net.SendCoins(ctxb, btcutil.SatoshiPerBitcoin, carol)
+	if err != nil {
+		t.Fatalf("unable to send coins to carol: %v", err)
+	}
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	chanPointCarol := openChannelAndAssert(ctxt, t, net, carol,
+		net.Alice, chanAmt, 0)
+
+	// Create invoices for Bob, which expect a payment from Carol for 1k
+	// satoshis with a different preimage each time.
+	const paymentAmt = 1000
+	for i := range testCases {
+		invoice := &lnrpc.Invoice{
+			Memo:  "testing",
+			Value: paymentAmt,
+		}
+		resp, err := net.Bob.AddInvoice(ctxb, invoice)
+		if err != nil {
+			t.Fatalf("unable to add invoice: %v", err)
+		}
+
+		testCases[i].hash = resp.RHash
+	}
+
+	// Wait for carol to recognize both the Channel from herself to Carol,
+	// and also the channel from Alice to Bob.
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	err = carol.WaitForNetworkChannelOpen(ctxt, chanPointCarol)
+	if err != nil {
+		t.Fatalf("carol didn't advertise her channel: %v", err)
+	}
+	ctxt, _ = context.WithTimeout(ctxb, timeout)
+	err = carol.WaitForNetworkChannelOpen(ctxt, chanPointAlice)
+	if err != nil {
+		t.Fatalf("carol didn't see the alice->bob channel before timeout: %v", err)
+	}
+
+	// Using Carol as the source, pay to the 5 invoices from Bob created above.
+	carolPayStream, err := carol.SendPayment(ctxb)
+	if err != nil {
+		t.Fatalf("unable to create payment stream for carol: %v", err)
+	}
+
+	// Concurrently pay off all 5 of Bob's invoices. Each of the goroutines
+	// will unblock on the recv once the HTLC it sent has been fully
+	// settled.
+	for _, test := range testCases {
+		sendReq := &lnrpc.SendRequest{
+			PaymentHash: test.hash,
+			Dest:        net.Bob.PubKey[:],
+			Amt:         paymentAmt,
+			MaxFee:      test.maxFee,
+		}
+
+		if err := carolPayStream.Send(sendReq); err != nil {
+			t.Fatalf("unable to send payment: %v", err)
+		}
+
+		if resp, err := carolPayStream.Recv(); err != nil {
+			t.Fatalf("payment stream has been closed: %v", err)
+		} else if resp.PaymentError != "" {
```

Comment:
This can be re-written a bit to make the flow of the test a bit more explicit: 
```go
if resp, err := carolPayStream.Recv(); err != nil {
    t.Fatalf("payment stream has been closed: %v", err)
}

switch {
    case resp.PaymentError != "" && test.expecteSucess:
        // fail the test
    case resp.PaymentError == "" && !test.expectSucess:
       // fail the test
}
```